File: package/ios/DataSource/HybridNativeListDataSource.swift
import DifferenceKit
import Foundation
import NitroModules

final class DiffableListItem: Differentiable {
    typealias DifferenceIdentifier = String

    let nativeItem: NativeListItem
    private let contentEqual: (NativeListItem, NativeListItem) -> Bool

    init(
        nativeItem: NativeListItem,
        contentEqual: @escaping (NativeListItem, NativeListItem) -> Bool
    ) {
        self.nativeItem = nativeItem
        self.contentEqual = contentEqual
    }

    var differenceIdentifier: String {
        return nativeItem.type + ":" + nativeItem.key
    }

    func isContentEqual(to source: DiffableListItem) -> Bool {
        if nativeItem.type != source.nativeItem.type {
            return false
        }
        if nativeItem.width != source.nativeItem.width {
            return false
        }
        if nativeItem.height != source.nativeItem.height {
            return false
        }
        return contentEqual(source.nativeItem, nativeItem)
    }
}

protocol NativeListDataSourceObserver: AnyObject {
    func dataSourceDidReload(
        _ dataSource: HybridNativeListDataSource,
        animated: Bool,
        changeset: StagedChangeset<[DiffableListItem]>?
    )
    func dataSourceDidInsert(_ dataSource: HybridNativeListDataSource, index: Int)
    func dataSourceDidUpdate(
        _ dataSource: HybridNativeListDataSource,
        index: Int,
        previousItem: NativeListItem
    )
    func dataSourceDidRemove(
        _ dataSource: HybridNativeListDataSource,
        index: Int,
        removedItem: NativeListItem
    )
    func dataSourceDidMove(_ dataSource: HybridNativeListDataSource, fromIndex: Int, toIndex: Int)
}

class HybridNativeListDataSource: HybridNativeListDataSourceSpec {
    weak var observer: NativeListDataSourceObserver?
    private var items: [DiffableListItem] = []
    private var pendingTargetItems: [DiffableListItem]?
    private var animatedReloadSourceItems: [DiffableListItem]?
    private var contentEqual: (NativeListItem, NativeListItem) -> Bool = { _, _ in false }

    func dispose() {
        observer = nil
        items.removeAll()
        pendingTargetItems = nil
        animatedReloadSourceItems = nil
        contentEqual = { _, _ in false }
    }

    func setContentEqualCallback(
        isContentEqual: @escaping (NativeListItem, NativeListItem) -> Bool
    ) throws {
        contentEqual = isContentEqual
    }

    func replaceData(items newItems: [NativeListItem], animated: Bool) throws {
        let targetItems = wrap(newItems)
        guard animated, observer != nil else {
            pendingTargetItems = nil
            animatedReloadSourceItems = nil
            items = targetItems
            observer?.dataSourceDidReload(self, animated: false, changeset: nil)
            return
        }

        animatedReloadSourceItems = items
        let changeset = StagedChangeset(source: items, target: targetItems)
        guard !changeset.isEmpty else {
            pendingTargetItems = nil
            animatedReloadSourceItems = nil
            items = targetItems
            observer?.dataSourceDidReload(self, animated: false, changeset: nil)
            return
        }

        pendingTargetItems = targetItems
        observer?.dataSourceDidReload(self, animated: true, changeset: changeset)
    }

    func insertItem(index: Double, item: NativeListItem) throws {
        let itemIndex = validInsertionIndex(index)
        let wrappedItem = wrap(item)
        items.insert(wrappedItem, at: itemIndex)
        observer?.dataSourceDidInsert(self, index: itemIndex)
    }

    func updateItem(index: Double, item: NativeListItem) throws {
        let itemIndex = validExistingIndex(index)
        let previousItem = items[itemIndex].nativeItem
        let wrappedItem = wrap(item)
        items[itemIndex] = wrappedItem
        observer?.dataSourceDidUpdate(self, index: itemIndex, previousItem: previousItem)
    }

    func removeItem(index: Double) throws {
        let itemIndex = validExistingIndex(index)
        let removedItem = items.remove(at: itemIndex).nativeItem
        observer?.dataSourceDidRemove(self, index: itemIndex, removedItem: removedItem)
    }

    func moveItem(fromIndex: Double, toIndex: Double) throws {
        let sourceIndex = validExistingIndex(fromIndex)
        let targetIndex = validExistingIndex(toIndex)
        let item = items.remove(at: sourceIndex)
        items.insert(item, at: targetIndex)
        observer?.dataSourceDidMove(self, fromIndex: sourceIndex, toIndex: targetIndex)
    }

    func getCount() throws -> Double {
        return Double(items.count)
    }

    func getItem(index: Double) throws -> NativeListItem {
        let itemIndex = validExistingIndex(index)
        return item(at: itemIndex)
    }

    func item(at index: Int) -> NativeListItem {
        return items[index].nativeItem
    }

    func itemForCollectionViewQuery(at index: Int) -> NativeListItem {
        if items.indices.contains(index) {
            return items[index].nativeItem
        }

        // During DifferenceKit staged reloads, UICollectionViewFlowLayout can still ask for
        // sizing information from the pre-animation snapshot after the stage data was applied.
        if let animatedReloadSourceItems, animatedReloadSourceItems.indices.contains(index) {
            return animatedReloadSourceItems[index].nativeItem
        }

        // Crash-resistance (patched): the UICollectionView prefetcher can query an index that no
        // longer exists when `items` is replaced faster than the collection view reconciles (e.g. the
        // app rebuilds the dataset as async data streams in). Crashing the whole app for a transient
        // off-by-one in a prefetch/sizing query is never acceptable, so clamp to the nearest valid
        // item instead of `preconditionFailure`. The next layout pass re-queries with a correct index.
        if let fallback = items.last?.nativeItem ?? animatedReloadSourceItems?.last?.nativeItem {
            return fallback
        }

        preconditionFailure("List item index \(index) is out of bounds for collection view query.")
    }

    func replaceWrappedItemsFromCollectionView(_ nextItems: [DiffableListItem]) {
        items = nextItems

        guard let pendingTargetItems else {
            animatedReloadSourceItems = nil
            return
        }

        if hasSameIdentity(nextItems, pendingTargetItems) {
            self.pendingTargetItems = nil
            animatedReloadSourceItems = nil
        }
    }

    func itemsForPremeasurement() -> [NativeListItem] {
        let sourceItems = pendingTargetItems ?? items
        return sourceItems.map { item in
            item.nativeItem
        }
    }

    private func wrap(_ item: NativeListItem) -> DiffableListItem {
        return DiffableListItem(nativeItem: item, contentEqual: contentEqual)
    }

    private func wrap(_ nativeItems: [NativeListItem]) -> [DiffableListItem] {
        return nativeItems.map { item in
            wrap(item)
        }
    }

    private func hasSameIdentity(
        _ firstItems: [DiffableListItem],
        _ secondItems: [DiffableListItem]
    ) -> Bool {
        if firstItems.count != secondItems.count {
            return false
        }

        return zip(firstItems, secondItems).allSatisfy { firstItem, secondItem in
            return firstItem.differenceIdentifier == secondItem.differenceIdentifier
        }
    }

    private func validExistingIndex(_ value: Double) -> Int {
        let index = Int(value)
        precondition(index >= 0 && index < items.count, "List index \(index) is out of bounds.")
        return index
    }

    private func validInsertionIndex(_ value: Double) -> Int {
        let index = Int(value)
        precondition(index >= 0 && index <= items.count, "List index \(index) is out of bounds.")
        return index
    }
}

File: package/android/src/main/java/com/margelo/nitro/reactnativelist/HybridNativeListDataSource.kt
package com.margelo.nitro.reactnativelist

import androidx.recyclerview.widget.DiffUtil

internal interface NativeListDataSourceObserver {
    fun dataSourceDidReload(diffResult: DiffUtil.DiffResult?, animated: Boolean)
    fun dataSourceDidInsert(index: Int)
    fun dataSourceDidUpdate(index: Int, previousItem: NativeListItem)
    fun dataSourceDidRemove(index: Int, removedItem: NativeListItem)
    fun dataSourceDidMove(fromIndex: Int, toIndex: Int)
}

class HybridNativeListDataSource : HybridNativeListDataSourceSpec() {
    internal var observer: NativeListDataSourceObserver? = null
    private var items: List<NativeListItem> = emptyList()
    private var isContentEqual: (oldItem: NativeListItem, newItem: NativeListItem) -> Boolean =
        { _, _ -> false }

    override fun dispose() {
        observer = null
        items = emptyList()
        isContentEqual = { _, _ -> false }
    }

    override fun setContentEqualCallback(
        isContentEqual: (oldItem: NativeListItem, newItem: NativeListItem) -> Boolean
    ) {
        this.isContentEqual = isContentEqual
    }

    override fun replaceData(items: Array<NativeListItem>, animated: Boolean) {
        val nextItems = items.toList()
        if (!animated) {
            this.items = nextItems
            observer?.dataSourceDidReload(null, false)
            return
        }

        val previousItems = this.items
        val callback = NativeDiffCallback(previousItems, nextItems, isContentEqual)
        val diffResult = DiffUtil.calculateDiff(callback, true)
        this.items = nextItems
        observer?.dataSourceDidReload(diffResult, true)
    }

    override fun insertItem(index: Double, item: NativeListItem) {
        val itemIndex = validateInsertionIndex(index.toInt())
        val mutableItems = items.toMutableList()
        mutableItems.add(itemIndex, item)
        items = mutableItems
        observer?.dataSourceDidInsert(itemIndex)
    }

    override fun updateItem(index: Double, item: NativeListItem) {
        val itemIndex = validateExistingIndex(index.toInt())
        val mutableItems = items.toMutableList()
        val previousItem = mutableItems[itemIndex]
        mutableItems[itemIndex] = item
        items = mutableItems
        observer?.dataSourceDidUpdate(itemIndex, previousItem)
    }

    override fun removeItem(index: Double) {
        val itemIndex = validateExistingIndex(index.toInt())
        val mutableItems = items.toMutableList()
        val removedItem = mutableItems.removeAt(itemIndex)
        items = mutableItems
        observer?.dataSourceDidRemove(itemIndex, removedItem)
    }

    override fun moveItem(fromIndex: Double, toIndex: Double) {
        val sourceIndex = validateExistingIndex(fromIndex.toInt())
        val targetIndex = validateExistingIndex(toIndex.toInt())
        val mutableItems = items.toMutableList()
        val item = mutableItems.removeAt(sourceIndex)
        mutableItems.add(targetIndex, item)
        items = mutableItems
        observer?.dataSourceDidMove(sourceIndex, targetIndex)
    }

    override fun getCount(): Double {
        return items.size.toDouble()
    }

    override fun getItem(index: Double): NativeListItem {
        val itemIndex = validateExistingIndex(index.toInt())
        return getItemAt(itemIndex)
    }

    internal fun getItemAt(index: Int): NativeListItem {
        return items[index]
    }

    internal fun getCountAsInt(): Int {
        return items.size
    }

    private fun validateExistingIndex(index: Int): Int {
        if (index < 0 || index >= items.size) {
            throw IndexOutOfBoundsException("List index $index is out of bounds.")
        }
        return index
    }

    private fun validateInsertionIndex(index: Int): Int {
        if (index < 0 || index > items.size) {
            throw IndexOutOfBoundsException("List index $index is out of bounds.")
        }
        return index
    }
}

private class NativeDiffCallback(
    private val oldItems: List<NativeListItem>,
    private val newItems: List<NativeListItem>,
    private val isContentEqual: (oldItem: NativeListItem, newItem: NativeListItem) -> Boolean
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldItems.size
    }

    override fun getNewListSize(): Int {
        return newItems.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldItems[oldItemPosition]
        val newItem = newItems[newItemPosition]
        return oldItem.key == newItem.key && oldItem.type == newItem.type
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldItems[oldItemPosition]
        val newItem = newItems[newItemPosition]

        if (oldItem.width != newItem.width) {
            return false
        }
        if (oldItem.height != newItem.height) {
            return false
        }

        return isContentEqual(oldItem, newItem)
    }
}
